Specify collections

NobodysNightmare opened the pull request:
Basic groundwork for https://community.openproject.org/work_packages/16986

This specification deals with collections, another quite fundamental concept that needs specification before it is too late to adapt existing implementations to the changes proposed.
# What are the goals?
- one specification to rule them all (currently collections are quite diverse regarding their implementation)
- different pagination options
  - no pagination
  - pagination where the client wants to display pages
  - pagination where the client wants it all (e.g. infinite scrolling)
- enough flexibility to allow extensions (e.g. links to filtered variants; read: in the **future**)
# What changes with regards to the existing API-Implementation?
- there is a set of required properties (currently collections look very different)
- `total` and `count` are now prefix-less (they are not very special at all)
- the `_type` of every collection is `Collection` instead of `pluralOfResourceName`
- the values are stored in `"_embedded": { "elements" }`, not `"_embedded": { "pluralOfResourceName" }`

NobodysNightmare committed:
restructure intro

- remove examples for simple object and collection object
- extra section for collections (to be specified)
- remove collection meta properties (they are not as cross-cutting as
the other meta properties, they are quite normal properties)

[ci skip]

NobodysNightmare committed:
basic collection specification

NobodysNightmare committed:
move collections into own group

add some example requests and responses

NobodysNightmare committed:
chose examples matching the response

NobodysNightmare committed:
fix cross ref

[ci skip]

NobodysNightmare committed:
improve wording

[ci skip]

NobodysNightmare committed:
collection-type: "collection" + more clarity

- more concise example-URLS
- explicitly mention "elements" property

[ci skip]

NobodysNightmare committed:
prepend /api/v3

- fewer possibilities to read the URL wrong (op.org/work_packages/1 is
already taken...)
- otherwise it would later become a pain to specify /api/v3 ^^

[ci skip]

TeatroIO commented:
I've prepared a stage. <a href='http://feature-specify-collections.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -229,6 +181,149 @@ If the *Formatable* is marked as **read only**, the `raw` attribute also becomes
+| jumpTo         | Templated link to jump to a specified offset                             | when offset based available |
+| nextOffset     | Link to retrieve the following page of elements (offset based)           | when offset based available |
+| previousOffset | Link to retrieve the preceding page of elements (offset based)           | when offset based available |
+| next           | Link to retrieve the elements following the current page (cursor based)  | when cursor based available |
```
If one of the methods (Offset) receives a suffix, the other method (cursor) should also receive one. Who knows, we might want to introduce a third method...

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -229,6 +181,149 @@ If the *Formatable* is marked as **read only**, the `raw` attribute also becomes
+but only a subset. In this case the client can issue further requests to retrieve the remaining elements.
+
+A collection carries meta information like the total count of elements in the collection or - in case of a paginated collection -
+the amount of elements returned in this response and action links to retrieve the remaining elements.
```
The three terms:
- pagination
- offset
- cursor 
  should be defined before the table of properties and links. Otherwise, the tables are hard to understand.

NobodysNightmare replied:
I did not really define those terms above, but I tried to move the text so that the reader at least roughly knows to what they will refer.

I'd like to keep the table above the example requests and I'd like to keep the examples with their explanations. That is why I had a hard time to move the actual definitions above the table.

ulferts commented:
:+1:

NobodysNightmare committed:
equal naming for offset and cursor

[ci skip]

NobodysNightmare committed:
move pagination intro above tables

give some context for words used in the table, real explanation is kept
below the tables

[ci skip]

NobodysNightmare committed:
split that line

[ci skip]

hschink commented on doc/apiv3-documentation.apib:
```diff
@@ -229,6 +181,148 @@ If the *Formatable* is marked as **read only**, the `raw` attribute also becomes
+                        "templated": true
+                    },
+                    "previousOffset": { "href": "/api/v3/examples?offset=0&pageSize=25" },
+                    "previous": { "href": "/api/v3/examples?before=bar&pageSize=25" }
```
I don't find a description for this link in the description. Might it be obsolete here?

NobodysNightmare replied:
absolutely!

NobodysNightmare committed:
update example links
